### DISCUSSION TITLE
Optimize my use of Dataset and DataLoader at inference

### DISCUSSION THREAD

**Discussion by pcanas** [Project-MONAI/MONAI#6700] [Q&A] [answered]
My team is currently using the following code snippet to generate a `monai.data.DataLoader` for inference:
```
        ds_p = monai.data.Dataset([deepcopy(image) for _ in range(5)], transform=get_tta_tfms())
        dl_p = monai.data.DataLoader(ds_p, batch_size=1, num_workers=1, shuffle=False)
```
Where `get_tta_tfms()` is a function that retrieves a Compose of monai transforms.

I would like to know if there is a better way of achieving the same result for optimal performance.

Thanks!

**Comment by KumoLiu** [comment]
Hi @pcanas, sorry for I didn't get your question. You can simply set `transform=Compose(...)` in the `Dataset`.
Thanks!

**Comment by pcanas** [comment]
No, that is not what I wanted to ask. My point is that creating a Dataset of 5 deep copied image just to later send them 1 by 1 into the DataLoader, it does not sound efficient. I remember with pytorch you could just use the `collate_fn` argument and achieve similar results.

My question is, how can I create this `DataLoader` in the most efficient way.

**Reply by KumoLiu** [reply]
Hi @pcanas, I think you can use `RepeatChanneld` and `SplitDimd` to achieve the same result.
https://github.com/Project-MONAI/MONAI/blob/6f8e63f6cb2932492610515cd9b62f581fcb6d3e/monai/transforms/utility/dictionary.py#L332
https://github.com/Project-MONAI/MONAI/blob/6f8e63f6cb2932492610515cd9b62f581fcb6d3e/monai/transforms/utility/dictionary.py#L382

Hope it can help you, thanks!

**Reply by pcanas** [reply]
Hi @KumoLiu thanks!
Could you provide a code example?

**Reply by KumoLiu** [reply, marked as answer]
Here is a simple demo.
```
trans = Compose([
         LoadImaged(...),
         EnsureChannelFirstd(...),
         RepeatChanneld(..., repeats=5),
         SplitDimd(..., dim=0, list_output=True)
])
ds_p = monai.data.Dataset([{"img": image_path}], transform=trans)
```

BTW, you can also specify your own `collate_fn` in MONAI. 
https://github.com/Project-MONAI/MONAI/blob/6f8e63f6cb2932492610515cd9b62f581fcb6d3e/monai/data/dataloader.py#L86-L87

**Reply by pcanas** [reply]
I am not sure whether this achieves the same results... This would create the same transformed image 5 different times, whereas the original code creates 5 different transformed images.

Am I right?

**Reply by KumoLiu** [reply]
You can add any other transforms for each key after `SplitDimd`(maybe with `list_output=False`), then you will get five different results.

**Reply by pcanas** [reply]
I see, thank you!!
